src/main/java/org/magicdgs/readtools/utils/read/stats/engine/ProperStatWindowCalculator.java

```diff
@@ -0,0 +1,426 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package org.magicdgs.readtools.utils.read.stats.engine;
+
+import org.magicdgs.readtools.utils.read.stats.PairEndReadStatFunction;
+import org.magicdgs.readtools.utils.read.stats.SingleReadStatFunction;
+import org.magicdgs.readtools.utils.read.stats.StatFunction;
+
+import com.google.common.annotations.VisibleForTesting;
+import htsjdk.samtools.util.CoordMath;
+import htsjdk.samtools.util.Locatable;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.codecs.table.TableFeature;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Calculator for several {@link StatFunction} over a window.
+ *
+ * <p>This calculator works only for {@link GATKRead} functions, concretely using implementations of
+ * {@link SingleReadStatFunction} and {@link PairEndReadStatFunction}.
+ *
+ * <p>Computation of statistics happens only for pair-end reads. By default, the calculator computes
+ * all the reads that are in the window ("total"), all proper pairs ("proper") and missing pairs in
+ * the file ("missing"). Values for {@link SingleReadStatFunction} are only computed for "proper"
+ * pairs; for {@link PairEndReadStatFunction} "missing" reads are not considered.
+ *
+ * @author Daniel Gomez-Sanchez (magicDGS)
+ */
+final class ProperStatWindowCalculator implements Locatable {
+
+    /** Column name for 'total' reads (first column in output). */
+    public static final String DEFAULT_TOTAL_COLUMN_NAME = "total";
+    /** Column name for 'proper' reads (second column in output). */
+    public static final String DEFAULT_PROPER_COLUMN_NAME = "proper";
+    /** Column name for 'missing' reads (third column in output). */
+    public static final String DEFAULT_MISSING_COLUMN_NAME = "missing";
+
+    // running statistics are always the same and should be in sync with the formatColumnNames/formatValues method
+    private static final List<String> DEFAULT_COLUMN_NAMES = Arrays.asList(
+            DEFAULT_TOTAL_COLUMN_NAME, DEFAULT_PROPER_COLUMN_NAME, DEFAULT_MISSING_COLUMN_NAME
+    );
+
+    // store the interval data
+    private final SimpleInterval window;
+
+    // store the current results (pairs)
+    private final Map<PairEndReadStatFunction, Object> pairEndResult;
+    // store the current results (single)
+    private final Map<SingleReadStatFunction, Object> singleResult;
+
+    // TODO: check if we can improve performance with a differen Map and a sensible cache value
+    // TODO: maybe a good idea is to use the PatriciaTrie<Map<PairEndReadStatFunction, Object>> instead
+    // TODO: an avoid the ProperReadHash
```

Comment:
This TODO is not longer valid - we need the `ProperReadHash` to store the information if the pair is in the window or before the window.

Reply:
Removed.